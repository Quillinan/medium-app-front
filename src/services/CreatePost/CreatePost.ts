import handleError from '@errors/HandleError/HandleError';
import { apiUrl, post } from '@services/Api/Api';
import { Post, ErrorResponse, CreatePostData } from '@utils/Types/Types';

export const createPost = async (
  postData: CreatePostData
): Promise<Post | ErrorResponse | undefined> => {
  const url = `${apiUrl}api/Post`;
  const formData = new FormData();

  formData.append('Title', postData.title);
  formData.append('Subtitle', postData.subtitle);
  formData.append('Content', postData.content);
  formData.append('AuthorId', postData.authorId);
  formData.append('AuthorName', postData.authorName);

  if (postData.coverImage) {
    formData.append('CoverImageData', postData.coverImage);
  }

  try {
    const response = await post(url, formData);
    return response as Post;
  } catch (error: any) {
    return handleError(error);
  }
};
