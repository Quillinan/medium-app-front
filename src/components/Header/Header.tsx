import Swal from 'sweetalert2';
import { CreatePostData } from '@utils/Types/Types';
import { createPost } from '@services/CreatePost/CreatePost';

interface HeaderProps {
  title: string;
  coverImage: File | null;
  setCoverImage: (file: File | null) => void;
  onTabChange: (tab: string) => void; // Adiciona a função onTabChange como prop
}

const Header: React.FC<HeaderProps> = ({
  title,
  coverImage,
  setCoverImage,
  onTabChange,
}) => {
  const createPostHandler = async () => {
    const draft = localStorage.getItem('draftPost');
    if (!draft) {
      console.error('Nenhum rascunho encontrado.');
      return;
    }

    const { title, subtitle, content } = JSON.parse(draft);
    const userId = sessionStorage.getItem('uniqueId') ?? 'erro';
    const username = sessionStorage.getItem('name') ?? 'erro';

    if (!userId || !username) {
      Swal.fire({
        title: 'Erro',
        text: 'UserId ou Username não encontrados.',
        icon: 'error',
      });
      return;
    }

    const postData: CreatePostData = {
      title,
      subtitle,
      content,
      authorId: userId,
      authorName: username,
      coverImage,
    };

    // Chama a função createPost do serviço
    const result = await createPost(postData);

    if (result) {
      Swal.fire({
        title: 'Sucesso',
        text: 'Post criado com sucesso!',
        icon: 'success',
      });
      localStorage.removeItem('draftPost'); // Limpa o rascunho
      setCoverImage(null); // Reseta a imagem de capa

      // Força a navegação para a aba 'Posts'
      onTabChange('Posts');
    } else {
      Swal.fire({
        title: 'Erro',
        text: 'Não foi possível criar o post. Tente novamente.',
        icon: 'error',
      });
    }
  };

  return (
    <header data-testid='header' className='bg-white shadow'>
      <div className='mx-auto max-w-7xl px-4 py-6 sm:px-6 lg:px-8 flex justify-between items-center'>
        <h1 className='text-3xl font-bold tracking-tight text-gray-900'>
          {title}
        </h1>
        {title === 'Criar Post' && (
          <button
            onClick={createPostHandler}
            className='bg-blue-500 text-white rounded px-4 py-1'
          >
            Publicar Post
          </button>
        )}
      </div>
    </header>
  );
};

export default Header;
