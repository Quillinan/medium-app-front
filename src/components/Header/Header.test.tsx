import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import Header from './Header';
import Swal from 'sweetalert2';
import { createPost } from '@services/CreatePost/CreatePost';

vi.mock('sweetalert2', () => ({
  default: {
    fire: vi.fn(),
  },
}));

vi.mock('@services/CreatePost/CreatePost', () => ({
  createPost: vi.fn(),
}));

describe('Header Component', () => {
  const mockSetCoverImage = vi.fn();
  const mockOnTabChange = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
    localStorage.clear();
    sessionStorage.clear();
  });

  it('should handle post creation and change tab after user confirms success', async () => {
    localStorage.setItem(
      'draftPost',
      JSON.stringify({
        title: 'Test Title',
        subtitle: 'Test Subtitle',
        content: 'Test Content',
      })
    );
    sessionStorage.setItem('uniqueId', 'mock-userId');
    sessionStorage.setItem('name', 'mock-username');
    (createPost as jest.Mock).mockResolvedValue(true);
    (Swal.fire as jest.Mock).mockResolvedValue({ isConfirmed: true });

    render(
      <Header
        title='Criar Post'
        coverImage={null}
        setCoverImage={mockSetCoverImage}
        onTabChange={mockOnTabChange}
      />
    );

    const buttonElement = screen.getByRole('button', {
      name: /Publicar Post/i,
    });
    fireEvent.click(buttonElement);

    await waitFor(() => {
      expect(Swal.fire).toHaveBeenCalledWith({
        title: 'Sucesso',
        text: 'Post criado com sucesso!',
        icon: 'success',
      });
    });

    expect(localStorage.getItem('draftPost')).toBeNull();
    expect(mockSetCoverImage).toHaveBeenCalledWith(null);
    expect(mockOnTabChange).toHaveBeenCalledWith('Posts');
  });

  it('should match snapshot', () => {
    const { asFragment } = render(
      <Header
        title='Criar Post'
        coverImage={null}
        setCoverImage={mockSetCoverImage}
        onTabChange={mockOnTabChange}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
