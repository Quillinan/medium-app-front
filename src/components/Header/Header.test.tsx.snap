// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render header name 1`] = `
<h1
  class="text-3xl font-bold tracking-tight text-gray-900"
>
  test
</h1>
`;
